driver/pdf2jpeg.man
.TH XScreenSaver 1 "07-Sep-2003 (4.13)" "X Version 11"
.SH NAME
pdf2jpeg - converts a PDF file to a JPEG file using Cocoa
.SH SYNOPSIS
.B pdf2jpeg
[\--verbose] [\--quality \fINN\fP] infile.pdf outfile.jpg
.SH DESCRIPTION
This reads a PDF file (for example, as written by the
.BR screencapture (1)
program) and writes a JPEG file.
.SH OPTIONS
.I pdf2jpeg
accepts the following options:
.TP 4
.B --verbose
Print diagnostics.
.TP 4
.B --quality \fINN\fP
JPEG compression factor.  Default 85%.
.SH BUGS
The input and output files must be files: pipes don't work.

This program is Cocoa-specific, so it won't work on non-MacOS systems.

This shouldn't need to be a part of the XScreenSaver distribution at
all, but Apple is COMPLETELY INSANE and made
.BR screencapture (1)
only write PDFs, with no simple way to convert that to something
less crazy.
.SH SEE ALSO
.BR screencapture (1),
.BR xscreensaver\-getimage\-desktop (1)
.SH COPYRIGHT
Copyright \(co 2003 by Jamie Zawinski.  Permission to use, copy,
modify, distribute, and sell this software and its documentation for
any purpose is hereby granted without fee, provided that the above
copyright notice appear in all copies and that both that copyright
notice and this permission notice appear in supporting documentation.
No representations are made about the suitability of this software for
any purpose.  It is provided "as is" without express or implied
warranty.
.SH AUTHOR
Jamie Zawinski <jwz@jwz.org>, 20-Oct-03.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


driver/pdf2jpeg.m
/* pdf2jpeg -- converts a PDF file to a JPEG file, using Cocoa
 *
 * Copyright (c) 2003, 2008 by Jamie Zawinski <jwz@jwz.org>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 *
 * Inspired by clues provided by Jan Kujawa and Jonathan Hendry.
 */

#import <Cocoa/Cocoa.h>
#include <stdio.h>
#include <stdlib.h>

int
main (int argc, char** argv)
{
  const char *progname = argv[0];
  const char *infile = 0, *outfile = 0;
  double compression = 0.85;
  double scale = 1.0;
  int verbose = 0;
  int i;

  for (i = 1; i < argc; i++)
    {
      char c;
      if (argv[i][0] == '-' && argv[i][1] == '-')
        argv[i]++;
      if (!strcmp (argv[i], "-q") ||
          !strcmp (argv[i], "-qual") ||
          !strcmp (argv[i], "-quality"))
        {
          int q;
          if (1 != sscanf (argv[++i], " %d %c", &q, &c) ||
              q < 5 || q > 100)
            {
              fprintf (stderr, "%s: quality must be 5 - 100 (%d)\n",
                       progname, q);
              goto USAGE;
            }
          compression = q / 100.0;
        }
      else if (!strcmp (argv[i], "-scale"))
        {
          float s;
          if (1 != sscanf (argv[++i], " %f %c", &s, &c) ||
              s <= 0 || s > 50)
            {
              fprintf (stderr, "%s: scale must be 0.0 - 50.0 (%f)\n",
                       progname, s);
              goto USAGE;
            }
          scale = s;
        }
      else if (!strcmp (argv[i], "-verbose"))
        verbose++;
      else if (!strcmp (argv[i], "-v") ||
               !strcmp (argv[i], "-vv") ||
               !strcmp (argv[i], "-vvv"))
        verbose += strlen(argv[i])-1;
      else if (argv[i][0] == '-')
        {
          fprintf (stderr, "%s: unknown option %s\n", progname, argv[i]);
          goto USAGE;
        }
      else if (!infile)
        infile  = argv[i];
      else if (!outfile)
        outfile = argv[i];
      else
        {
        USAGE:
          fprintf (stderr,
                   "usage: %s [-verbose] [-scale N] [-quality NN] "
                   "infile.pdf outfile.jpg\n",
                   progname);
          exit (1);
        }
    }

  if (!infile || !outfile)
    goto USAGE;


  // Much of Cocoa needs one of these to be available.
  NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

  //Need an NSApp instance to make [NSImage TIFFRepresentation] work
  NSApp = [NSApplication sharedApplication];
  [NSApp autorelease];

  if (verbose)
    fprintf (stderr, "%s: reading %s...\n", progname, infile);

  // Load the PDF file into an NSData object:
  NSData *pdf_data = [NSData dataWithContentsOfFile:
                               [NSString stringWithCString:infile
                                         encoding:NSUTF8StringEncoding]];

  // Create an NSPDFImageRep from the data:
  NSPDFImageRep *pdf_rep = [NSPDFImageRep imageRepWithData:pdf_data];

  // Create an NSImage instance
  NSRect rect;
  rect.size = [pdf_rep size];
  rect.size.width  *= scale;
  rect.size.height *= scale;
  rect.origin.x = rect.origin.y = 0;
  NSImage *image = [[NSImage alloc] initWithSize:rect.size];

  // Draw the PDFImageRep in the NSImage
  [image lockFocus];
  [pdf_rep drawInRect:rect];
  [image unlockFocus];

  // Load the NSImage's contents into an NSBitmapImageRep:
  NSBitmapImageRep *bit_rep = [NSBitmapImageRep
                                imageRepWithData:[image TIFFRepresentation]];

  // Write the bitmapImageRep to a JPEG file:
  if (bit_rep == nil)
    {
      fprintf (stderr, "%s: error converting image?\n", argv[0]);
      exit (1);
    }

  if (verbose)
    fprintf (stderr, "%s: writing %s (%d%% quality)...\n",
             progname, outfile, (int) (compression * 100));

  NSDictionary *props = [NSDictionary
                          dictionaryWithObject:
                            [NSNumber numberWithFloat:compression]
                          forKey:NSImageCompressionFactor];
  NSData *jpeg_data = [bit_rep representationUsingType:NSJPEGFileType
                               properties:props];

  [jpeg_data writeToFile:
               [NSString stringWithCString:outfile
                         encoding:NSUTF8StringEncoding]
             atomically:YES];
  [image release];

  [pool release];
  exit (0);
}
